utests/test_cudafft.c

```diff
@@ -0,0 +1,100 @@
+/* Copyright 2019. Christian Holme.
+ * All rights reserved. Use of this source code is governed by
+ * a BSD-style license which can be found in the LICENSE file.
+ *
+ * Authors:
+ * 2019 Christian Holme <christian.holme@med.uni-goettingen.de>
+ */
+
+
+
+
+#include <complex.h>
+#include <assert.h>
+
+#include "num/fft.h"
+#include "num/rand.h"
+
+#include "num/multind.h"
+#include "num/flpmath.h"
+
+#include "misc/debug.h"
+#include "misc/misc.h"
+
+#include "utest.h"
+
+enum { N = 6 };
+
+
+
+static bool run_cuda_fft_test (const long* dims, const unsigned int flags,
+			       const complex float* in, complex float* cpu_inout,
+			       complex float* gpu_inout, complex float* gpu_result)
+{
+	md_copy(N, dims, cpu_inout, in, CFL_SIZE);
+	md_copy(N, dims, gpu_inout, in, CFL_SIZE);
+
+	// create fft plan
+	const struct operator_s* fftplan = fft_create(N, dims, flags, cpu_inout, cpu_inout, false);
+
+	fft_exec(fftplan, cpu_inout, cpu_inout);
+	fft_exec(fftplan, gpu_inout, gpu_inout);
+
+	md_copy(N, dims, gpu_result, gpu_inout, CFL_SIZE);
+
+	UT_ASSERT(md_znrmse(N, dims, cpu_inout, gpu_result) < UT_TOL);
+
+	fft_free(fftplan);
+}
+
+
+
+
+static bool test_cuda_fft (void)
+{
+#ifdef USE_CUDA
+	// TODO: detect if GPU works
+
+	num_rand_init(5);
+
+	const long dims[N] = { 4, 4, 4, 4, 4, 4};
+	const long transform_dims[][N] = {
+						{1, 1, 1, 0, 0, 0},
+						{1, 1, 0, 0, 1, 0},
+						{1, 0, 1, 0, 1, 0},
+						{1, 1, 0, 1, 1, 1},
+					};
+
+
+	complex float* in = md_alloc(N, dims, CFL_SIZE);
+	md_gaussian_rand(N, dims, in);
+
+	complex float* cpu_inout = md_alloc(N, dims, CFL_SIZE);
+	complex float* gpu_inout = md_alloc_gpu(N, dims, CFL_SIZE);
+	complex float* gpu_result = md_alloc(N, dims, CFL_SIZE);
+
+	for (unsigned int i = 0; i < (sizeof(transform_dims)/sizeof(*transform_dims)); ++i) {
+
+		unsigned int flags = 0;
+		const long* tdims = &(transform_dims[i][0]);
+		for (int j = 0; j < N; ++j)
+			if (tdims[j])
+				flags = MD_SET(flags, j);
+
+		run_cuda_fft_test (dims, flags, in, cpu_inout, gpu_inout, gpu_result);
+	}
+	md_free(gpu_result);
+	md_free(gpu_inout);
+	md_free(cpu_inout);
+	md_free(in);
+
+	return true;
+
+#else
+	// cannot test cuFFT without CUDA
+	return true;
```

Comment:
but you can compile everything and you should. Also there needs to be a global flag/targt for running gpu tests I think. Otherwise this will be far too annoying.